added for review

ik8 opened the pull request.

ik8 committed:
added for review

ik8 commented on Board.java:
```diff
@@ -0,0 +1,521 @@
+
+    public Board(Game g,int NumofHoles)
+    {
+        Currx = 0; Curry = 0;
```
save as const, no need for it to be in the constructor

ik8 commented on Board.java:
```diff
@@ -0,0 +1,521 @@
+        setBackground(Color.black);
+        parent = g;
+
+        BorderLayout map = new BorderLayout();
```
all init can be done without the constructor

ik8 commented on Board.java:
```diff
@@ -0,0 +1,521 @@
+    {
+        int i = e.getKeyCode();
+        
+        if((i == KeyEvent.VK_1)&&(!board[Currx][Curry].isPlaced))
```
all the if statements can be a one function that uses a dictionary of the cases

ik8 commented on Board.java:
```diff
@@ -0,0 +1,521 @@
+
+        for(int i = 0 ; i < 3 ; i++)
+        {
+            g.drawLine(i * 60 * 3 - 5, 0, i * 60 * 3 - 5, 530);
```
magic numbers

ik8 commented on Board.java:
```diff
@@ -0,0 +1,521 @@
+
+    public synchronized void paintComponent(Graphics g)
+	{
+		super.paintComponent(g);
```
redundant tab

ik8 commented on Board.java:
```diff
@@ -0,0 +1,521 @@
+    public static int RandomNum()
+    {
+        Random r = new Random();
+        int num = r.nextInt(9);
```
const

ik8 commented on Menu.java:
```diff
@@ -0,0 +1,86 @@
+        Menu m = new Menu();
+    }
+
+    public Menu()
```
can be simplified to function that gets a JRadioButton and list of actions to apply to it

ik8 commented on Game.java:
```diff
@@ -0,0 +1,76 @@
+    int difficulty;
+    AutoSolver t;
+
+    public Game(int difficulty)
```
I see it happens a lot so I'll write it one time.
1) A Java function should be a maximum of 30-40 lines, your function can be split to 3 or 4 small ones and the code will look a lot better
2) Try using const or maybe use enums(The best!)
3) There is a Java convention that you type '{' after class/function/if/else without the line break, It is not an ugly C# language.
4) I think you should try using generics like List<int>, I find it more convenient.
5) You have a lot of for loops, there is a beautiful thing since Java 8 called **Stream**, check it out
6) Try reading to code conventions for Java for prettier code
7) Use small classes and function unless they have a lot of logic(you don't have it here)

Beside that nice job! I love it that you keep learning and develop your skills.

AmirBurdini committed:
july 20 update

same square identical numbers fixed
added pause to solver thread

ik8 commented on Cell.java:
```diff
@@ -0,0 +1,57 @@
+    public Cell(Board p, int x, int y){
+        
+        num = 0; 
+        Posibilities = new boolean[] {true,true,true,true,true,true,true,true,true};
```
Generate a true * x somehow, this is ugly

AmirBurdini replied:
done

ik8 commented on Cell.java:
```diff
@@ -0,0 +1,57 @@
+    int y;
+    
+    // constructor, empty Cell
+    public Cell(Board p, int x, int y){
```
Why you need the board?

AmirBurdini replied:
fixed

ik8 commented on Cell.java:
```diff
@@ -0,0 +1,57 @@
+        {
+            if(Posibilities[i])
+            {
+                return false;
```
Maybe using a dictionary is better than iterating n^2 every time(In the worst case)

AmirBurdini replied:
done

ik8 commented on DEF.java:
```diff
@@ -0,0 +1,23 @@
+public class DEF {
```
Create enum with values

AmirBurdini replied:
instead of DEF class? or inside of it?

ik8 replied:
Instead, no need for a class only for const values

ik8 commented on Game.java:
```diff
@@ -0,0 +1,109 @@
+    int solverSpeed; // how fast the solver is displayed
+    boolean buttonPressed;
+
+    public Game(int difficulty) {
```
Difficulity should be enum

AmirBurdini replied:
done

ik8 commented on Game.java:
```diff
@@ -0,0 +1,109 @@
+        setFocusable(true);
+        setVisible(true);
+        setSize(DEF.GAME_HEIGHT, DEF.GAME_WIDTH);
+        setLocation(410, 105);
```
No consts for the location?

AmirBurdini replied:
done

ik8 commented on Game.java:
```diff
@@ -0,0 +1,109 @@
+        public void stateChanged(ChangeEvent e) {
+
+            JSlider x = (JSlider) e.getSource();
+            DEF.SOLVER_VELOCITY = x.getValue();
```
Why change const value? You shouldn't do that

AmirBurdini replied:
not a const but a static var, used to determine all possible solvers speeds

ik8 replied:
So move it from def to another location, def should contain only cons values(definitions which you indicate at the start of your program and will not change)

ik8 commented on Menu.java:
```diff
@@ -0,0 +1,99 @@
+        JRadioButton easy, medium, hard, insane, help;
+        ButtonGroup difficultyOptions = new ButtonGroup();
+
+        easy = new JRadioButton("easy");
```
Can be rewrited with a loop for all levels

AmirBurdini committed:
modifications 29th july 2020 14:48

-handeled key events using a hashtable to minimize runtime
-added consts for screen location
-added enum difficulty instead of using numerical values
-removed non final var from the DEF class
-removed unused var and argument in the Cell.java class

*not a complete edition*
